Foldseek driver + Structure base Pangenome

metehaansever opened the pull request.

metehaansever committed:
Update structure pan

metehaansever committed:
Fix process problem and file check

metehaansever committed:
Update Structure pan base on args

metehaansever committed:
Update weight_dir

metehaansever committed:
Add anvi-run-foldseek

metehaansever committed:
Update anvi-run-foldseek to anvi-structural-pan-genome

metehaansever committed:
Fixy for parameters

metehaansever committed:
merge master

metehaansever committed:
fixy

metehaansever committed:
Add get_gene_clusters_based_on_structure func

Add --mode structure params for anvi-display-pan

metehaansever committed:
Add mode params

metehaansever committed:
Add mode param

metehaansever committed:
Add process and get_foldseek_results method

metehaansever committed:
Add foldseek changes

metehaansever committed:
Add mode to anvi-pan-genome

metehaansever committed:
Remove mode in anvi-display-pan

metehaansever committed:
Update metavar

metehaansever committed:
fixy

metehaansever committed:
better group for mode

metehaansever committed:
Update default params in foldseek driver

metehaansever committed:
Add output_file_path to init

FlorianTrigodet committed:
fix result directory

Commit:
foldseek db name fix

metehaansever committed:
Add prostt5-weight-dir

metehaansever committed:
Update naming

metehaansever committed:
Update naming

metehaansever committed:
Remove redundant codes

metehaansever committed:
Add anvi-setup-prostt5

metehaansever committed:
Add new parameter for weight dir

metehaansever committed:
Fix downloading of prostt5 issue

metehaansever committed:
Add prostt5-weight-dir option

metehaansever commented:
I need to update md file for anvi-setup-prostt5 and in get_gene_clusters_de_novo we can update the name of blastall_results variable, which is not cover blast, diamond and foldseek.

metehaansever committed:
Update docs for prostt5

FlorianTrigodet commented:
Coupe of notes to discuss:
- I think we don't need structurepan.py. It only contains a class to download the ProstT5 weights using a foldseek command. And ProstT5 itself is, and will not be specific to structure pangenome in the future. Since the download is managed by a foldseek command line, it could be migrated to the foldseek driver. Or somewhere else if you have a better opinion.
- ProstT5 is written with a combination of upper and lowercase. This is not the most important note, but we should write it properly when we can (with the exception of anvi-setup-prostt5 because I can see that all the command line in anvi'o are lower case - with minor exceptions).

metehaansever committed:
Fix prostt5-weight-dir param

meren commented:
@FlorianTrigodet's note on`structurepan.py` is spot on.

I will leave some additional comments on the code directly :)

I can't wait to test it. Florian, we will have to create a minimal test case for this so it can be run by our nightly action scripts.

meren commented on anvio/__init__.py:
```diff
@@ -3669,7 +3669,27 @@ def TABULATE(table, header, numalign="right", max_width=0):
                      "and provide an absolutely ridiculous string after it (we suggest 'cattywampus', but you do you)."}
-    )
+    ),
+    'mode': (
```
Please change the new community parameter to `pan-mode` so we don't run into trouble in the long run.

meren commented on anvio/__init__.py:
```diff
@@ -3669,7 +3669,27 @@ def TABULATE(table, header, numalign="right", max_width=0):
+         'type': str,
+         'help': 'Use this flag to set mode to structure or sequence.'}
+    ),
+   'prostt5-weight-dir': (
```
Why not call it `prostt5-data-dir`?

Indentations of new entries in `__init__` do not much each other.

meren replied:
please don't forget to change the help files with the new parameter name too.

metehaansever replied:
Frankly it seems more correct to use _prostt5-weight-dir_ here. Because what we downloaded is a model's weight file and Foldseek uses it with the same name, for example:` foldseek createdb db.fasta db --prostt5-model weights`

meren replied:
It doesn't matter though. As far as our users are concerned, it is yet another data they are downloading. Regardless what individual data items mean to individual processes, we can have a common vocabulary we bring in from the outside world. What we download from COGs is FASTA files. What we download from KEGG is models. We call all of them data on anvi'o side.

meren commented on anvio/__init__.py:
```diff
@@ -3669,7 +3669,27 @@ def TABULATE(table, header, numalign="right", max_width=0):
+    'mode': (
+        ['--mode', '-M'],
+        {'default': None,
+         'metavar': 'structure',
```
`metavar` value should be `CHOICE` so it looks good in the help menu.

meren commented on anvio/panops.py:
```diff
@@ -89,12 +90,18 @@ def __init__(self, args=None, run=run, progress=progress):
             raise ConfigError("Please set a project name using --project-name or -n.")
 
+        if self.de_novo_compute_mode == 'structure':
+            self.skip_alignments = True
```
this condition should also set `self.skip_homogeneity` to `True`.

meren commented on anvio/panops.py:
```diff
@@ -89,12 +90,18 @@ def __init__(self, args=None, run=run, progress=progress):
         self.enforce_hierarchical_clustering = A('enforce_hierarchical_clustering')
         self.enforce_the_analysis_of_excessive_number_of_genomes = anvio.USER_KNOWS_IT_IS_NOT_A_GOOD_IDEA
 
+        self.de_novo_compute_mode = A('mode') or 'sequence'
```
somewhere in `check_params` function we should check for the use of `--pan-mode structure` together with `--user-defined-gene-clusters` parameter (and prevent it basically).

meren reviewed (commented):
Some quick suggestions

metehaansever committed:
Update Prostt5setupWeight dir

metehaansever committed:
Delete structurepan.py

metehaansever committed:
Update driver with prostt5 class

meren commented:
This is the error you get when you don't have foldseek:

```
Config Error: An anvi'o function needs 'foldseek' to be installed on your system, but it
              doesn't seem to appear in your path :/ If you are certain you have it on your
              system (for instance you can run it by typing 'foldseek' in your terminal
              window), you may want to send a detailed bug report. Sorry!
```

It should be more instructive and help the user how they can get the foldseek installed.

meren commented:
And of course this PR should be accompanied by another PR for anvio.org to update the installation instructions so the users who install anvio-dev have all the necessary tools from the get go.

meren commented:
When you run the following command,

```
anvi-pan-genome -g TEST-GENOMES.db -n TEST --mode structure --force-overwrite
```

Without setting up PROSTT5 first, you get the following error:

```
File/Path Error: No such file: '/Users/meren/github/anvio/anvio/data/misc/PROSTT5/weights' :/
```

Obviously this is not the most useful error. What do we expect the user to do with this information? There should be an earlier check in anvi'o style of exception-driven programming and things shouldn't even come all the way down here before we tell the user "you seem to be interested in doing this, but this other stuff is not in place yet -- please run this program first blah blah".

meren commented:
I haven't looked at the code yet, but this is not how we generally download things in anvi'o:

<img width="445" alt="image" src="https://github.com/user-attachments/assets/2d9a81cb-5ecd-48c5-881b-74b3cc547ca4">

If you look at KEGG or COG downloads as examples you will see that we strive to report a percentage and ETA so the users are not forced to looking at a stale screen without knowing whether something is actually happening or not.

metehaansever commented:
> When you run the following command,
> 
> ```
> anvi-pan-genome -g TEST-GENOMES.db -n TEST --mode structure --force-overwrite
> ```
> 
> Without setting up PROSTT5 first, you get the following error:
> 
> ```
> File/Path Error: No such file: '/Users/meren/github/anvio/anvio/data/misc/PROSTT5/weights' :/
> ```
> 
> Obviously this is not the most useful error. What do we expect the user to do with this information? There should be an earlier check in anvi'o style of exception-driven programming and things shouldn't even come all the way down here before we tell the user "you seem to be interested in doing this, but this other stuff is not in place yet -- please run this program first blah blah".

There is a problem with error messages in general, yes I will overhaul all of them. :thinking:

metehaansever committed:
Update mode to pan-mode

metehaansever committed:
Add choice_of_pangenome to constant

metehaansever committed:
Update mode to pan-mode

metehaansever committed:
Update mode

metehaansever committed:
Add user_defined_gene_clusters control

metehaansever committed:
Update prostt5-weight-dir to data-dir

metehaansever committed:
Fixy for user_defined_gene_cluster and pan-mode structure

metehaansever committed:
Update error message

metehaansever committed:
fixy

metehaansever committed:
fixy for ProstT5 naming

metehaansever committed:
Add get_gene_clusters_structure

metehaansever committed:
Add classical pangenome for structural pangenome

metehaansever committed:
Changes from meeting

metehaansever committed:
fixy

meren committed:
get gene cluster representatives now works

meren committed:
generate FASTA file for foldseek analysis

meren committed:
cosmetics

FlorianTrigodet commented:
We will need 2 additional layers in the structure based pan, with information taken from the AA pangenome. 
One layer will be "AA-GC Count" for the number of amino acid GCs that ended up in a structure GC.
The second one would be a barplot (stacked bars) called "AA-GC Type" with three information: "core", "singleton", "else". You would count how many genes in the structure GC came from a AA core or singleton GC. If non of them, then you count it as "else".

meren committed:
Merge branch 'master' into foldseek-driver

meren committed:
critical updates to get things to run

we now go with `protein_structure_informed_gene_clusters_dict`
tentatively as per the discussions in #lab.

once we have everything ready, meren will go through all the variable
names to try to fix the ones that are not quite working, and we will
open this up to the community input.

metehaansever committed:
Update PSGC dict

meren committed:
new param: --foldseek-search-results

meren committed:
use existing search results when provided

meren committed:
fix typo

metehaansever committed:
Fix file path of foldseek db

metehaansever committed:
fixy

metehaansever committed:
Merge branch 'foldseek-driver' of https://github.com/merenlab/anvio into foldseek-driver

metehaansever committed:
Add functions for PSGC

metehaansever committed:
fixy

metehaansever committed:
fixy

metehaansever committed:
Add self.output_file

metehaansever committed:
Fix gene_functions empty state

meren committed:
de_novo_compute_mode -> pan_mode

meren committed:
update meta values with mode

meren committed:
new table to track gc <-> psgc associations

this will be useful also to generate many of the additional data items
to understand the nature of psgcs

meren committed:
TableForPSGCGCAssociations table class

meren committed:
dict building tracks psgc <-> gc associations

I also removed a risky if statement that may mask de novo gene clusters
missing in protein structure-informed gene clusters

meren committed:
keep user up to date with what is going on

meren committed:
populate TableForPSGCGCAssociations

meren committed:
oops. TableForPSGCGCAssociations table -> dbops

meren committed:
Merge branch 'foldseek-driver' of github.com:merenlab/anvio into foldseek-driver

metehaansever committed:
typo

metehaansever committed:
Remove functions in PSGC  data

metehaansever committed:
Add num GCs in PSGC as additional layer

metehaansever committed:
Add num genes and gcs in PSGC

metehaansever committed:
Add default names of psgc

metehaansever committed:
add constants of core|singleton|else

metehaansever committed:
add all additional layers

This commit introduces new visualization layers for PSGCs (Protein Structure Gene
Clusters) that classify genes into three categories:
- Core genes (present in all genomes)
- Singleton genes (present in only one genome)
- Other genes (present in some but not all genomes)

metehaansever committed:
fixy

metehaansever committed:
Merge branch 'master' into foldseek-driver

metehaansever committed:
Add de novo gene type classification to PSGCs

metehaansever committed:
Fix naming

metehaansever committed:
fixy

metehaansever committed:
fixy :)

meren committed:
Merge branch 'master' into foldseek-driver

meren committed:
on a second thought, probably better to allow

meren commented:
There is something annoying going on here and we need to fix it soon. As you can see here, SCGs (the red layer) are not coming together nicely anymore.

As the flat branches in the dendrogram like the one I marked below has an alphabetical order, when we named our gene clusters based on their frequency across genomes in the regular pangenomics workflow, this would enforce SCGs to be next to each other nicely in the core. But it seems it is not happening anymore because the way we name them does not care about their frequencies across genomes :/ I'm putting it here as a note so we take care of it at some point.

![image](https://github.com/user-attachments/assets/694b60a5-461f-48a8-9e43-d6b4e6398994)

metehaansever commented:
We should add pan-db-type variant as 'structure' to get structural specific data table in interactive class when pan-mode == 'structure'.

meren committed:
better tracking of modes and terms

meren committed:
better mode tracking in panops

meren committed:
link mode to db-variant for posterity

meren committed:
no more `pan_mode` variable in self

meren committed:
ensure the file is not there prior

meren committed:
cosmetics

meren committed:
a new pan_gc_tracker table

The purpose of this table will be to keep the de novo gene clusters in
the pan-db so we can go back to their contents when we need to

meren committed:
populate de novo gc tracker table

meren committed:
TablesForGeneClusters can handle default and tracker modes

meren committed:
tables specific to db-variant structure-informed

metehaansever committed:
Add GC info for each gene

metehaansever committed:
remove console.log

metehaansever committed:
Add GC info popover

metehaansever committed:
Add no gene lenghts overflow in GC popover

meren committed:
fixy

meren committed:
update pan-db version

meren committed:
pan-db migration script from v21 to v22

metehaansever committed:
add gc_tracker and gc_psgc_associations to interactive

metehaansever committed:
Update gene entries by gene_caller_id

metehaansever committed:
Add get_psgc_data endpoint

metehaansever committed:
Update gene cluster table on inspection page

metehaansever committed:
Merge branch 'foldseek-driver' of https://github.com/merenlab/anvio into foldseek-driver

meren committed:
Merge branch 'master' into foldseek-driver

metehaansever committed:
Update NGGC and NgGC float to int

metehaansever committed:
Fix gcSequence width and wrap bug

meren committed:
store gc_psgc_associations only in STRUCTURE_MODE

meren commented:
@metehaansever, https://github.com/merenlab/anvio/pull/2364/commits/3084b8f41d2eb69b3c6d7911d1f21a921173b9ed should fix the bug above. Please test :)

metehaansever committed:
Add get_psgc_type_data

Gets GC type data for inspection page.

metehaansever committed:
Add gc type data to item_additional_data table

metehaansever committed:
Add GC type letter to inspection page

metehaansever committed:
Add color to GC type

metehaansever committed:
fixy

metehaansever committed:
Add gc type info into GC table

metehaansever committed:
fixy

metehaansever committed:
Fix for resizing related console bug

metehaansever committed:
Fix psgc_composition display table bug

metehaansever committed:
Update search table

metehaansever committed:
Fix performance on inspection

metehaansever committed:
Add gene types in PSGC into gc_types table

metehaansever committed:
Fix gc_types search problem

metehaansever committed:
Update search result table with json parser

metehaansever committed:
Fix performance issue on larger content

metehaansever committed:
fix sequence drawing bottleneck

metehaansever committed:
Update minimize dom manipulation

metehaansever committed:
Add tspan element to dom once

metehaansever committed:
Fix make geneclusters page looks same for firefox

metehaansever committed:
Add bg to gc_type table

metehaansever committed:
Add search btn css

metehaansever committed:
Add search functionality

metehaansever committed:
Add search elements

metehaansever committed:
Add additional control for classical pangenome

meren committed:
Merge branch 'master' into foldseek-driver

metehaansever committed:
Update buttons to bs4 buttons

metehaansever committed:
Add psgc_data length check

metehaansever committed:
Update genome filtering with include option

metehaansever committed:
Update naming

metehaansever committed:
Update color settings buttons

metehaansever committed:
Update color coding of genes

metehaansever committed:
Add foldseek citation

metehaansever committed:
Add structure-informed-pangenomics into test

metehaansever committed:
Add structure-informed-pangenomics script

metehaansever committed:
Update dir name

metehaansever committed:
fixy

metehaansever committed:
Revert gene colors

metehaansever committed:
Update prostt5 doc

metehaansever committed:
Remove homogeneity check in structure-informed-pangenomics script

metehaansever committed:
Create mock data for component test

metehaansever committed:
Update component test with collection

meren committed:
psgc_composition -> GCs in PSCG

metehaansever committed:
Merge branch 'master' into foldseek-driver

metehaansever committed:
Update mouse_hover_table to parse Json

meren committed:
some cosmetics in the summary output

metehaansever committed:
Add navbar opacity

metehaansever committed:
Add structure-informed summary

- gc_tracker
- gc_psgc_associations
Tables are added to Pan summary.

metehaansever committed:
Update data type to list

metehaansever committed:
Add gc_id info

metehaansever committed:
Prettify

metehaansever committed:
Remove num_genes_in_psgc

metehaansever committed:
Update func params

metehaansever committed:
Update add_layers_to_view params

metehaansever committed:
Remove redundant code

metehaansever committed:
fixy

metehaansever committed:
Add PSGC info into Gene Cluster bin modal

metehaansever committed:
Fix Gene Cluster table overflow behavior

metehaansever committed:
Update GC tables col size

metehaansever committed:
Add additional control for mode structure

metehaansever committed:
fixy

metehaansever committed:
Add temp mode

metehaansever committed:
Update bin mode with structure

metehaansever committed:
Update GC bins mode

metehaansever committed:
Typo

metehaansever committed:
Fix Layer Order

metehaansever committed:
Revert

metehaansever committed:
Add init-pan-mode for anvi-summarize

metehaansever committed:
Add new args for structure-pan

metehaansever committed:
Add gc_psgc_associations data into gz summary

metehaansever committed:
Add gene cluster matches info into .gz

metehaansever committed:
remove console.log

metehaansever committed:
Remove redundant data in summary

metehaansever committed:
Fix accurate naming for additional layer

metehaansever committed:
Update data structure

Remove genes related data from gc types data

metehaansever committed:
Update init-pan-mode check for classic pan

metehaansever committed:
fixy

metehaansever committed:
Add structure-mode

- add structure mode on client side instead of passing in interactive.py
- Update Structure Pan related ui elements such as mode and search text

metehaansever committed:
add header id

meren committed:
Merge branch 'master' into foldseek-driver

required some tiny adjustements

meren committed:
Merge branch 'master' into foldseek-driver

FlorianTrigodet committed:
Functions to compare pan dbs

FlorianTrigodet committed:
Improve speed of compare_pan

FlorianTrigodet committed:
fix init_function bug in compare_pan

FlorianTrigodet committed:
Merge branch 'master' into foldseek-driver

FlorianTrigodet committed:
typo

FlorianTrigodet committed:
New arg for compared pan

FlorianTrigodet committed:
moved compare_pan_to to ComparePan class

FlorianTrigodet committed:
Unwanted spaces :(

FlorianTrigodet committed:
comparepan output txt

FlorianTrigodet committed:
Simplify comparpan main dict

FlorianTrigodet committed:
Report no-function as None

FlorianTrigodet committed:
migrate anvi-setup-prostt5

FlorianTrigodet committed:
Stop comparing pan if no differences

FlorianTrigodet committed:
sanity check for same number of genes in two pan

FlorianTrigodet committed:
Merge branch 'master' into foldseek-driver

FlorianTrigodet committed:
Fix compare pan add info to item additional

FlorianTrigodet committed:
Fix - classic summary will work now

FlorianTrigodet committed:
ComparePan handles missing genes

FlorianTrigodet committed:
Wish we could display add layers in groups

FlorianTrigodet committed:
Choose a better representative sequence

FlorianTrigodet committed:
Fix AA alignment wrapping issue

FlorianTrigodet committed:
Fix issue to identify identical gene clusters

FlorianTrigodet committed:
Representative sequences should have NO GAPS

FlorianTrigodet committed:
Better AAI calculation (without double gaps)

FlorianTrigodet committed:
Fixy

FlorianTrigodet committed:
New AAI_min_nonzero in pangenome

FlorianTrigodet committed:
Merge branch 'master' into foldseek-driver

FlorianTrigodet committed:
Catching error

FlorianTrigodet committed:
Update ComparePan

FlorianTrigodet committed:
structural pangenome now can output a conv pangenome database

FlorianTrigodet committed:
Associated flag for anvi-pan-genome to skip conv pan

FlorianTrigodet committed:
New program anvi-compare-pan

FlorianTrigodet committed:
New flag now in effect for --skip-sequence-pan

FlorianTrigodet committed:
Merge branch 'master' into foldseek-driver

FlorianTrigodet committed:
New item data group for compare pan

FlorianTrigodet committed:
All PSCG-GC comparison are part of default ComparePan

FlorianTrigodet committed:
Fix bugs in ComparePan

FlorianTrigodet committed:
Rework the ComparePan

FlorianTrigodet committed:
New color scheme and new evenness score layer

FlorianTrigodet committed:
Merge branch 'master' into foldseek-driver

FlorianTrigodet committed:
Add structures-txt artifact; refactor external-structures as a specialization

FlorianTrigodet committed:
predicted-structure input path + TM-score output

FlorianTrigodet committed:
user-supplied structures + TM-score-driven clustering

FlorianTrigodet committed:
Retire anvi-setup-prostt5 from the user side

FlorianTrigodet committed:
Tests for the structures-txt path

FlorianTrigodet committed:
Merge branch 'master' into foldseek-driver

FlorianTrigodet committed:
anvi-split works with struct-pan

FlorianTrigodet committed:
move the code for picking GC representatives + flag to export them

FlorianTrigodet committed:
remove mention of structure

FlorianTrigodet committed:
fix ruff issues

FlorianTrigodet committed:
fully remove prostt5

FlorianTrigodet committed:
fix KeyError in ComparePan function summary

meren committed:
Merge branch 'master' into foldseek-driver
